design/cluster-api-provider-metal3/node_reuse.md

```diff
@@ -0,0 +1,250 @@
+to be able to reuse the same pool of hosts during re-provisioning (i.e. when
+BareMetalHost is going through deprovisioning -> provisioning). However, there
+is no mechanism currently available in Metal³ to pick the same pool of hosts
+that was released while upgrading/remediation - for the next provisioning phase.
+
+**NOTE:** This proposal focuses on upgrade use cases but it isn't limited to
+only upgrade. Other use cases like remediation can also benefit from this
+feature as well.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I would like the same nodes to be used when they go for
+reprovisioning (i.e. deprovisioning -> provisioning) during the upgrade operation.
+
+#### Story 2
+
+As a cluster admin, I would like the same nodes to be used when they go for
+reprovisioning (i.e. deprovisioning -> provisioning) during the remediation
+operation.
+
+#### Story 3
+
+As a cluster admin, I would like any available (i.e. state: `Ready`) BMH to be
+selected randomly for a new CAPI Machine -> Metal3Machine(M3M) when nodeReuse is
+False.
+
+### Goals
+
+Add a logic to mark, filter and pick those marked BMH(s) when going through a
+re-provisioning cycle as part of the upgrade procedure.
+
+### Non-Goals
+
+Support node reuse for Machines which are created independently of Kubeadm
+control plane controller/Machine Deployment(KCP/MD). Currently we are not trying
+to support node reuse feature for the Machines not owned by higher level objects
+like KCP/MD. Because, if there is no KCP/MD that Machine could point to, then
+Cluster API Provider Metal3 (CAPM3) Machine controller will fail to first, set
+the `consumerRef` in spec of the BMH(s) and second, to filter the BMH(s) based
+on the `consumerRef`.
+
+## Proposal
+
+We would like to propose an interface to mark BMH(s) that we want to reuse after
+deprovisioning and second, add a selection/filtering mechanism in CAPM3 Machine
+controller that will pick those BMH(s) with a specific mark (to be exact, based
+on the value of `consumerRef`).
+
+## Design Details
+
+Normally, before deprovisioning the host, the controller removes the M3M name
+from the`consumerRef` field of the BMH when M3M is deleted.
+
+In case of nodeReuse field is set to True, to make sure that a BMH(s) is not
+utilized by other M3M(s) after it has been deprovisioned, the `consumerRef`
+field will store the name of KCP/MD even when it is in `Ready` state. During the
+upgrade, after KCP/MD creates new Machines, CAPM3 will create new M3M(s) and will
+filter out BMH(s) that have the name of KCP/MD in the `consumerRef` field. If the
+name of the KCP/MD matches the name set in `spec.consumerRef.name` field of BMH,
+then CAPM3 Machine controller will associate that BMH to the M3M.
+
+As such, even after deprovisioning we will keep the `consumerRef` on a BMH and
+when M3M will be requesting a new BMH, CAPM3 will filter BMH(s) based on their
+consumerRef and will pick up those that match the name of KCP/MD.
```

Comment:
Maintaining state like this in consumerRef doesn't seem ideal, we already have https://github.com/metal3-io/cluster-api-provider-metal3/blob/master/api/v1alpha4/metal3machine_types.go#L49 that could be used to filter BMH resources based on some label - would that be sufficient to address this requirement?

Reply:
We can rely on label/annotation, and we had a thought about using annotation first. But we thought relying on labels/annotation is not the recommended to manage a behavior. As such, we tried to avoid tuning the functionality via labels/annotation.

Reply:
I agree that we can just leave a label on a BMH during de-provisioning, and pick that up while next provisioning by relying on `HostSelector` to find it out. TBH, I don't have a strong opinion about which option would be the ideal. But if you think labeling is good enough in our case, then we can update the proposal. 

WDYT @maelk

Reply:
Setting the consumer ref would effectively reserve that BMH, by showing that it is currently consumed by the KCP / MD object. I think that this is the goal of that field of the API, so I am not sure why we should add a separate API to achieve the same thing.

Reply:
I tend to agree to @hardys comment, as it seems more in line with the management of similar problems in other contexts, for example [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/). ConsumerRef may be seen more as a technical details and implicit, whereas the usage of a label could be more clear and explicit as a public API (to indicate the desire of reuse)

Reply:
Hi @andfasano  @hardys ,
@furkatgofurov7 has now updated the proposal to use a new label `infrastructure.cluster.x-k8s.io/node-reuse` rather than consumerRef. However, @maelk  pointed out that we can't use HostSelector, but need to add something similar due to the following:

When nodeReuse is true, and if the user wants to scale out KCP/MD, then HostSelector will try to find a BMH with `infrastructure.cluster.x-k8s.io/node-reuse`. However, it is obvious that there won't be any BMH available with the label since it is scale out, so scale won't happen. That's why we need to add some sort of similar logic as HostSelector, which would be responsible to filter out BMHs only for non scale-out use cases. When it is scale out, we still would rely on HostSelector, as we do it now.

Does it make sense to you all ?